src/lib/cost_model/abstract_cost_feature_proxy.cpp

```diff
@@ -0,0 +1,67 @@
+#include "abstract_cost_feature_proxy.hpp"
+
+#include <cmath>
+
+#include "logical_query_plan/abstract_lqp_node.hpp"
+
+namespace opossum {
+
+CostFeatureVariant AbstractCostFeatureProxy::extract_feature(const CostFeature cost_feature) const {
+  switch (cost_feature) {
+    case CostFeature::InputRowCountProduct:
+      return {extract_feature(CostFeature::LeftInputRowCount).scalar() *
+              extract_feature(CostFeature::RightInputRowCount).scalar()};
+    case CostFeature::LeftInputReferenceRowCount:
+      return extract_feature(CostFeature::LeftInputIsReferences).boolean()
+                 ? extract_feature(CostFeature::LeftInputRowCount).scalar()
+                 : 0.0f;
```

Comment:
Yes! Because when you're retrieving `LeftInputReferenceRowCount` you're not implying that the left input actually is a ReferenceColumn. You're just saying "if it is one, give me it's row count, otherwise zero, so I can multiply it with penalty weight to account for the de-referencing that we have to do."

Reply:
If someone would imply assume that the column is of reference type he might just multiply the cost with 0.0f?

Reply:
not sure we're speaking about the same thing. The cost model will have a "penalty weight" for de-referencing a value. That makes it very easy to describe the cost model using just data and not using any "if reference input then ..." logic.

Look at this, e.g.:

```c++
+  config.table_scan_models[CostModelRuntimeTableScanType::ColumnValueNumeric] =
+      CostFeatureWeights{{CostFeature::LeftInputReferenceRowCount, 0.151131599208f},
+                         {CostFeature::LeftInputRowCount, 0.15565695035f},
+                         {CostFeature::OutputRowCount, 0.112082904016f},
+                         {CostFeature::OutputDereferenceRowCount, 0.0248249284018f}};
```

The model has a cost for each input row as well as an additional cost for each row if it is a reference. All clear?